File: src/main/java/uk/gov/justice/digital/hmpps/oauth2server/security/UserDetailsImpl.java
package uk.gov.justice.digital.hmpps.oauth2server.security;

import lombok.Data;
import lombok.EqualsAndHashCode;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.userdetails.User;

import java.util.Collection;

@SuppressWarnings("serial")
@EqualsAndHashCode(callSuper = true)
@Data
public class UserDetailsImpl extends User implements UserPersonDetails {
    private final String name;
    private final String firstName;
    private final String authSource;

    UserDetailsImpl(final String username, final String name, final String password,
                    final boolean enabled, final boolean accountNonExpired, final boolean credentialsNonExpired,
                    final boolean accountNonLocked, final Collection<? extends GrantedAuthority> authorities, final String authSource) {
        super(username, password, enabled, accountNonExpired, credentialsNonExpired, accountNonLocked, authorities);
        this.name = name;
        this.firstName = name;
        this.authSource = authSource;
    }

    UserDetailsImpl(final String username, final String name, final Collection<GrantedAuthority> authorities, final String authSource) {
        super(username, "", authorities);
        this.name = name;
        this.firstName = name;
        this.authSource = authSource;
    }

    @Override
    public boolean isAdmin() {
        return false;
    }
}

File: src/main/java/uk/gov/justice/digital/hmpps/oauth2server/security/UserDetailsImpl.kt
package uk.gov.justice.digital.hmpps.oauth2server.security

import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.userdetails.User

class UserDetailsImpl(username: String, private val name: String, authorities: Collection<GrantedAuthority>,
                      private val authSource: String, private val userId: String, val jwtId: String) :
    User(username, "", authorities), UserPersonDetails {
  override fun getName(): String = name
  override fun getAuthSource(): String = authSource
  override fun getFirstName(): String = name
  override fun getUserId(): String = userId

  override fun isAdmin(): Boolean = false

  override fun toUser(): uk.gov.justice.digital.hmpps.oauth2server.auth.model.User {
    throw IllegalStateException("Can't be converted into user")
  }

  companion object {
    // This class is serialized to the database (oauth_code table) during /auth/oauth/authorize and then read back
    // during /auth/oauth/token.  Therefore implemented serial version UID, although breaking changes should increment.
    private const val serialVersionUID = 1L
  }
}
